portal/src/graphql/portal/CreateOAuthClientScreen.tsx

```diff
@@ -0,0 +1,250 @@
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react";
+import {
+  Callout,
+  Dialog,
+  DialogFooter,
+  DirectionalHint,
+  IconButton,
+  Label,
+  PrimaryButton,
+  Text,
+} from "@fluentui/react";
+import { useNavigate, useParams } from "react-router-dom";
+import produce from "immer";
+import deepEqual from "deep-equal";
+import { FormattedMessage } from "@oursky/react-messageformat";
+
+import ShowError from "../../ShowError";
+import ShowLoading from "../../ShowLoading";
+import ModifyOAuthClientForm from "./ModifyOAuthClientForm";
+import ButtonWithLoading from "../../ButtonWithLoading";
+import NavigationBlockerDialog from "../../NavigationBlockerDialog";
+import NavBreadcrumb, { BreadcrumbItem } from "../../NavBreadcrumb";
+import { useAppConfigQuery } from "./query/appConfigQuery";
+import { useUpdateAppConfigMutation } from "./mutations/updateAppConfigMutation";
+import { OAuthClientConfig, PortalAPIAppConfig } from "../../types";
+import { clearEmptyObject } from "../../util/misc";
+import { genRandomHexadecimalString } from "../../util/random";
+import { copyToClipboard } from "../../util/clipboard";
+
+import styles from "./CreateOAuthClientScreen.module.scss";
+
+interface CreateOAuthClientFormProps {
+  rawAppConfig: PortalAPIAppConfig;
+}
+
+interface CreateClientSuccessDialogProps {
+  visible: boolean;
+  clientId: string;
+}
+const CreateClientSuccessDialog: React.FC<CreateClientSuccessDialogProps> = function CreateClientSuccessDialog(
+  props: CreateClientSuccessDialogProps
+) {
+  const { visible, clientId } = props;
+  const navigate = useNavigate();
+
+  const calloutIntervalRef = useRef<number>();
```

Comment:
Why use `useRef` instead of `useState`?

Reply:
This is because `calloutIntervalRef` is only used for cleanup, UI does not depend on this variable

Reply:
As it is now, there could be multiple active timeout interval, causing glitchy UI when tapping copy button repeatedly. Also, I think dismissing the callout after a timeout is counted as a UI concern.

I'd prefer using the `isCalloutVisible` variable to drive `setTimeout` side effect and remove the `useRef`, so the timeout effect is contained entirely within a single useEffect call:
```ts
const [calloutActiveCount, setCalloutActiveCount] = useState(0);
useEffect(() => {
    if (calloutActiveCount === 0) {
        return;
    }

    setIsCalloutVisible(true);
    const handle = setTimeout(() => setIsCalloutVisible(false), 3000);
    return () => clearTimeout(handle);
}, [calloutActiveCount]);

const onCopyClick = useCallback(() => {
    copyToClipboard(clientId);
    setCalloutActiveCount(c => c + 1);
}, [clientId]);
```

Optionally, you can also set the `key` prop of `Callout` to be `calloutActiveCount` to replay the animation on tap.

Reply:
I see, when repeatedly click copy, the callout will be set to invisible by first timeout, the code above fixes the problem